store/sqlstore/migration.go

```diff
@@ -0,0 +1,194 @@
+	BackoffTime int // Time to wait between retries
+	NumRetries  int // Number of retries in case of failed migration
+}
+
+func NewMigrationRunner(s SqlStore, o MigrationOptions) *MigrationRunner {
+	if o.LockTimeout == 0 {
+		o.LockTimeout = MigrationDefaultLockTimeout
+	}
+	if o.BackoffTime == 0 {
+		o.BackoffTime = MigrationDefaultBackoffTime
+	}
+	if o.NumRetries == 0 {
+		o.NumRetries = MigrationDefaultNumRetries
+	}
+	return &MigrationRunner{
+		supplier: s,
+		done:     make(chan struct{}),
+		options:  o,
+	}
+}
+
+// Add checks if the migration should be executed and adds it to queue
+func (r *MigrationRunner) Add(m AsyncMigration) error {
+	// check status in Systems table
+	currentStatus, appErr := r.supplier.System().GetByName("migration_" + m.Name())
+	if appErr == nil {
+		if currentStatus.Value == string(AsyncMigrationStatusComplete) || currentStatus.Value == string(AsyncMigrationStatusSkip) {
+			return nil
+		}
+	}
+	// get status from migration
+	status, err := m.GetStatus(r.supplier)
+	if err != nil {
+		return err
+	}
+	if status == AsyncMigrationStatusComplete || status == AsyncMigrationStatusSkip {
+		return nil
+	}
+	r.migrations = append(r.migrations, m)
+	return nil
+}
+
+// Run all queued migrations sequentially
+func (r *MigrationRunner) Run() error {
+	go func() {
+		ctx, cancelFn := context.WithCancel(context.Background())
+		defer cancelFn()
+		defer close(r.done)
+		r.cancelFn = cancelFn
+
+		for _, m := range r.migrations {
+			// function that will try to execute migration
+			migrate := func() error {
+				var err error
+				conn, err := createConnectionWithLockTimeout(ctx, r.supplier, r.options.LockTimeout)
+				if err != nil {
+					mlog.Error("Failed to setup connection", mlog.Err(err))
+					return err
+				}
+				defer releaseConnection(ctx, r.supplier, conn)
+
+				// run migration
+				status, err := m.Execute(ctx, r.supplier, conn)
+				if err != nil {
+					mlog.Error("Failed to execute migration", mlog.Err(err))
+					return err
+				}
+				if status == AsyncMigrationStatusComplete || status == AsyncMigrationStatusSkip {
+					// save migration status
+					r.supplier.System().SaveOrUpdate(&model.System{Name: "migration_" + m.Name(), Value: string(status)})
+				} else if status == AsyncMigrationStatusFailed {
+					mlog.Error("Failed migration " + m.Name())
+					return errors.New("Failed migration")
+				} else {
+					// should we return error here to retry?
+					return errors.New("Invalid result from migration")
+				}
+				return nil
+			}
+			// retry migration if it fails
+			for i := 0; i < r.options.NumRetries; i++ {
+				err := migrate()
+				if err == nil {
+					break
+				}
+				select {
+				case <-ctx.Done():
+					return
+				// wait before trying again
+				case <-time.After(time.Duration(r.options.BackoffTime) * time.Second):
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+// Wait returns after all migrations are processed (executed, skipped or failed)
```

Comment:
We should also add a `WaitWithTimeout` where it waits for a certain amount of time for migrations to be over, and if it does not finish, it cancels it. Then add it to the shutdown code where the application is shutting down. This will be useful to ensure we gracefully shutdown any active async migrations if the app is suddenly shut down.